File: server/samples/burp/BurpAiBridgePanel.java
package server.samples.burp;

import burp.IBurpExtender;
import burp.IBurpExtenderCallbacks;
import burp.IContextMenuFactory;
import burp.IContextMenuInvocation;
import burp.IHttpRequestResponse;
import burp.IRequestInfo;
import burp.ITab;

import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JMenuItem;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTabbedPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import java.awt.BorderLayout;
import java.awt.Font;
import java.awt.GridLayout;
import java.io.BufferedReader;
import java.io.OutputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/*
 * Burp AI Bridge companion panel sketch for Burp Suite Pro.
 * This is a starter template that shows the panel actions and HTTP calls.
 */
public class BurpAiBridgePanel implements IBurpExtender, ITab, IContextMenuFactory {
    private IBurpExtenderCallbacks callbacks;
    private JPanel panel;
    private JTextArea output;
    private JTextArea requestView;
    private JTextArea responseView;
    private JTextField baseUrlField;
    private JTextField issueField;
    private JTextField snapshotField;
    private JTextField currentModelField;
    private JTextField switchModelField;
    private IHttpRequestResponse[] selectedMessages = new IHttpRequestResponse[0];
    private String baseUrl = "http://127.0.0.1:8000";
    private String currentIssueId = "";
    private String currentSnapshotId = "";
    private String currentTargetUrl = "";
    private String currentModel = "";

    @Override
    public void registerExtenderCallbacks(IBurpExtenderCallbacks callbacks) {
        this.callbacks = callbacks;
        callbacks.setExtensionName("Burp AI Bridge Companion");
        callbacks.registerContextMenuFactory(this);

        panel = new JPanel(new BorderLayout());
        output = new JTextArea(24, 100);
        output.setEditable(false);
        output.setLineWrap(false);
        output.setWrapStyleWord(false);
        output.setFont(new Font("Consolas", Font.PLAIN, 12));
        requestView = new JTextArea(18, 100);
        requestView.setEditable(false);
        requestView.setLineWrap(false);
        requestView.setWrapStyleWord(false);
        requestView.setFont(new Font("Consolas", Font.PLAIN, 12));
        responseView = new JTextArea(18, 100);
        responseView.setEditable(false);
        responseView.setLineWrap(false);
        responseView.setWrapStyleWord(false);
        responseView.setFont(new Font("Consolas", Font.PLAIN, 12));

        JPanel controls = new JPanel(new GridLayout(2, 1));
        JPanel toolbar = new JPanel();
        JPanel info = new JPanel();
        baseUrlField = new JTextField(baseUrl, 28);
        issueField = new JTextField("", 18);
        snapshotField = new JTextField("", 18);
        currentModelField = new JTextField("", 18);
        switchModelField = new JTextField("", 18);
        issueField.setEditable(false);
        snapshotField.setEditable(false);
        currentModelField.setEditable(false);
        JButton openPlan = new JButton("Open Plan in Repeater");
        JButton syncTabs = new JButton("Sync Selected Tabs");
        JButton refresh = new JButton("Refresh Panel State");
        JButton recommendMoves = new JButton("Show Burp Moves");
        JButton refreshModels = new JButton("Refresh Model Options");
        JButton switchModel = new JButton("Switch Active Model");
        JButton resetModel = new JButton("Reset Model");

        openPlan.addActionListener(event -> openPlanInRepeater());
        syncTabs.addActionListener(event -> syncSelectedTabs());
        refresh.addActionListener(event -> refreshPanelState());
        recommendMoves.addActionListener(event -> showBurpMoves());
        refreshModels.addActionListener(event -> refreshModelOptions());
        switchModel.addActionListener(event -> switchActiveModel());
        resetModel.addActionListener(event -> resetActiveModel());

        toolbar.add(new JLabel("Bridge URL"));
        toolbar.add(baseUrlField);
        toolbar.add(openPlan);
        toolbar.add(syncTabs);
        toolbar.add(refresh);
        toolbar.add(recommendMoves);
        toolbar.add(refreshModels);
        info.add(new JLabel("Issue"));
        info.add(issueField);
        info.add(new JLabel("Snapshot"));
        info.add(snapshotField);
        info.add(new JLabel("Model"));
        info.add(currentModelField);
        info.add(new JLabel("Switch To"));
        info.add(switchModelField);
        info.add(switchModel);
        info.add(resetModel);

        controls.add(toolbar);
        controls.add(info);

        panel.add(controls, BorderLayout.NORTH);
        JScrollPane outputScroll = new JScrollPane(output);
        outputScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        JScrollPane requestScroll = new JScrollPane(requestView);
        requestScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        JScrollPane responseScroll = new JScrollPane(responseView);
        responseScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        JTabbedPane tabs = new JTabbedPane();
        tabs.addTab("Bridge Output", outputScroll);
        tabs.addTab("Selected Request", requestScroll);
        tabs.addTab("Selected Response", responseScroll);
        panel.add(tabs, BorderLayout.CENTER);

        callbacks.addSuiteTab(this);
        append("Burp AI Bridge companion loaded.\n");
    }

    @Override
    public String getTabCaption() {
        return "AI Bridge";
    }

    @Override
    public java.awt.Component getUiComponent() {
        return panel;
    }

    @Override
    public List<JMenuItem> createMenuItems(IContextMenuInvocation invocation) {
        selectedMessages = invocation.getSelectedMessages() != null ? invocation.getSelectedMessages() : new IHttpRequestResponse[0];
        List<JMenuItem> items = new ArrayList<>();
        JMenuItem openItem = new JMenuItem("AI Bridge: Open plan in Repeater");
        JMenuItem syncItem = new JMenuItem("AI Bridge: Sync selected tabs");
        JMenuItem refreshItem = new JMenuItem("AI Bridge: Refresh panel state");
        JMenuItem recommendItem = new JMenuItem("AI Bridge: Show Burp moves");
        openItem.addActionListener(event -> openPlanInRepeater());
        syncItem.addActionListener(event -> syncSelectedTabs());
        refreshItem.addActionListener(event -> refreshPanelState());
        recommendItem.addActionListener(event -> showBurpMoves());
        items.add(openItem);
        items.add(syncItem);
        items.add(refreshItem);
        items.add(recommendItem);
        return items;
    }

    private void openPlanInRepeater() {
        String payload = buildSingleMessagePayload("scanner");
        if (payload == null) {
            append("No selected Burp message for plan open.\n");
            return;
        }
        String result = postJson("/api/burp/open-repeater-plan?dispatch=true", payload);
        syncIdentity(result);
        append("Open plan result\n" + result + "\n\n");
    }

    private void syncSelectedTabs() {
        String payload = buildSyncPayload();
        if (payload == null) {
            append("Need one baseline plus one or more selected Repeater tabs.\n");
            return;
        }
        String result = postJson("/api/burp/repeater-sync", payload);
        syncIdentity(result);
        append("Repeater sync result\n" + result + "\n\n");
    }

    private void refreshPanelState() {
        String payload = buildSingleMessagePayload("repeater");
        if (payload == null) {
            payload = "{\"raw_request\":\"\",\"target_url\":\"" + escapeJson(currentTargetUrl) + "\",\"use_burp_mcp_context\":true}";
        }
        StringBuilder path = new StringBuilder("/api/burp/panel-state");
        List<String> params = new ArrayList<>();
        if (!currentIssueId.isEmpty()) {
            params.add("issue_id=" + currentIssueId);
        }
        if (!currentSnapshotId.isEmpty()) {
            params.add("snapshot_id=" + currentSnapshotId);
        }
        if (!params.isEmpty()) {
            path.append("?").append(String.join("&", params));
        }
        String result = postJson(path.toString(), payload);
        syncIdentity(result);
        append("Panel state\n" + result + "\n\n");
    }

    private void showBurpMoves() {
        String payload = buildSingleMessagePayload("scanner");
        if (payload == null) {
            payload = "{\"raw_request\":\"\",\"target_url\":\"" + escapeJson(currentTargetUrl) + "\",\"use_burp_mcp_context\":true}";
        }
        String result = postJson("/api/burp/capability-recommendations", payload);
        append("Burp capability recommendations\n" + result + "\n\n");
    }

    private void refreshModelOptions() {
        String result = getJson("/api/runtime/model-options");
        syncModelState(result);
        append("Model options refreshed.\n");
    }

    private void switchActiveModel() {
        String modelName = switchModelField.getText().trim();
        if (modelName.isEmpty()) {
            append("Enter a model tag before switching.\n");
            return;
        }
        String payload = "{\"model_name\":\"" + escapeJson(modelName) + "\",\"source\":\"burp-panel\"}";
        String result = postJson("/api/runtime/model-selection", payload);
        syncModelState(result);
        append("Active model switched.\n");
    }

    private void resetActiveModel() {
        String result = deleteJson("/api/runtime/model-selection");
        syncModelState(result);
        append("Active model reset to env default.\n");
    }

    private String buildSingleMessagePayload(String sourceTool) {
        if (selectedMessages.length == 0) {
            return null;
        }
        IHttpRequestResponse message = selectedMessages[0];
        IRequestInfo requestInfo = callbacks.getHelpers().analyzeRequest(message);
        String requestText = callbacks.getHelpers().bytesToString(message.getRequest());
        String responseText = callbacks.getHelpers().bytesToString(message.getResponse());
        requestView.setText(requestText);
        requestView.setCaretPosition(0);
        responseView.setText(responseText);
        responseView.setCaretPosition(0);
        String targetUrl = requestInfo.getUrl() != null ? requestInfo.getUrl().toString() : "";
        currentTargetUrl = targetUrl.isEmpty() ? currentTargetUrl : targetUrl;
        return "{"
            + "\"raw_request\":\"" + escapeJson(requestText) + "\","
            + "\"raw_response\":\"" + escapeJson(responseText) + "\","
            + "\"target_url\":\"" + escapeJson(targetUrl) + "\","
            + "\"http_method\":\"" + escapeJson(requestInfo.getMethod()) + "\","
            + "\"source_tool\":\"" + escapeJson(sourceTool) + "\","
            + "\"use_burp_mcp_context\":true"
            + "}";
    }

    private String buildSyncPayload() {
        if (selectedMessages.length == 0) {
            return null;
        }
        String base = buildSingleMessagePayload("repeater");
        String baselineResponse = callbacks.getHelpers().bytesToString(selectedMessages[0].getResponse());
        String baselineRequest = callbacks.getHelpers().bytesToString(selectedMessages[0].getRequest());
        requestView.setText(baselineRequest);
        requestView.setCaretPosition(0);
        responseView.setText(baselineResponse);
        responseView.setCaretPosition(0);
        StringBuilder observations = new StringBuilder("[");
        for (int index = 1; index < selectedMessages.length; index++) {
            IHttpRequestResponse message = selectedMessages[index];
            IRequestInfo info = callbacks.getHelpers().analyzeRequest(message);
            if (index > 1) {
                observations.append(",");
            }
            observations.append("{")
                .append("\"tab_name\":\"selected-tab-").append(index).append("\",")
                .append("\"request_text\":\"").append(escapeJson(callbacks.getHelpers().bytesToString(message.getRequest()))).append("\",")
                .append("\"response_text\":\"").append(escapeJson(callbacks.getHelpers().bytesToString(message.getResponse()))).append("\",")
                .append("\"request_ref\":\"").append(escapeJson(info.getUrl() != null ? info.getUrl().toString() : "")).append("\"")
                .append("}");
        }
        observations.append("]");
        return base.substring(0, base.length() - 1)
            + ",\"baseline_response_text\":\"" + escapeJson(baselineResponse) + "\""
            + ",\"repeater_variant_observations\":" + observations
            + "}";
    }

    private String postJson(String path, String body) {
        try {
            URL url = new URL(baseUrl + path);
            baseUrl = baseUrlField.getText().trim().isEmpty() ? baseUrl : baseUrlField.getText().trim();
            url = new URL(baseUrl + path);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            try (OutputStream outputStream = connection.getOutputStream()) {
                outputStream.write(body.getBytes(StandardCharsets.UTF_8));
            }
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                StringBuilder result = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    result.append(line).append("\n");
                }
                return result.toString();
            }
        } catch (Exception exception) {
            return "{\"error\":\"" + escapeJson(exception.toString()) + "\"}";
        }
    }

    private String getJson(String path) {
        try {
            baseUrl = baseUrlField.getText().trim().isEmpty() ? baseUrl : baseUrlField.getText().trim();
            URL url = new URL(baseUrl + path);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                StringBuilder result = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    result.append(line).append("\n");
                }
                return result.toString();
            }
        } catch (Exception exception) {
            return "{\"error\":\"" + escapeJson(exception.toString()) + "\"}";
        }
    }

    private String deleteJson(String path) {
        try {
            baseUrl = baseUrlField.getText().trim().isEmpty() ? baseUrl : baseUrlField.getText().trim();
            URL url = new URL(baseUrl + path);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("DELETE");
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                StringBuilder result = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    result.append(line).append("\n");
                }
                return result.toString();
            }
        } catch (Exception exception) {
            return "{\"error\":\"" + escapeJson(exception.toString()) + "\"}";
        }
    }

    private void append(String text) {
        output.append(text);
    }

    private void syncIdentity(String body) {
        currentIssueId = extractJsonValue(body, "issue_id", currentIssueId);
        currentSnapshotId = extractJsonValue(body, "snapshot_id", currentSnapshotId);
        currentTargetUrl = extractJsonValue(body, "target_url", currentTargetUrl);
        issueField.setText(currentIssueId);
        snapshotField.setText(currentSnapshotId);
    }

    private void syncModelState(String body) {
        currentModel = extractJsonValue(body, "current_model", currentModel);
        currentModelField.setText(currentModel);
        if (switchModelField.getText().trim().isEmpty()) {
            switchModelField.setText(extractJsonValue(body, "deep", currentModel));
        }
    }

    private String extractJsonValue(String body, String key, String fallback) {
        String marker = "\"" + key + "\":";
        int start = body.indexOf(marker);
        if (start < 0) {
            return fallback;
        }
        int quote = body.indexOf("\"", start + marker.length());
        int end = quote >= 0 ? body.indexOf("\"", quote + 1) : -1;
        if (quote >= 0 && end > quote) {
            return body.substring(quote + 1, end);
        }
        return fallback;
    }

    private String escapeJson(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\r", "\\r").replace("\n", "\\n");
    }
}

File: server/samples/burp/BurpAiBridgePanel.py
#
# Burp AI Bridge companion panel sketch for Burp Suite Pro + Jython 2.7.
# This is a practical starter template, not a packaged production extension.
#

from burp import IBurpExtender
from burp import ITab
from burp import IContextMenuFactory
from java.awt import BorderLayout
from java.awt import GridLayout
from java.awt import Font
from java.util import ArrayList
from javax.swing import JButton
from javax.swing import JLabel
from javax.swing import JMenuItem
from javax.swing import JPanel
from javax.swing import JScrollPane
from javax.swing import JTabbedPane
from javax.swing import JTextArea
from javax.swing import JTextField
from javax.swing import ScrollPaneConstants
from javax.swing import SwingUtilities
import json
import urllib2


class BurpExtender(IBurpExtender, ITab, IContextMenuFactory):
    def registerExtenderCallbacks(self, callbacks):
        self._callbacks = callbacks
        self._helpers = callbacks.getHelpers()
        self._base_url = "http://127.0.0.1:8000"
        self._selected_messages = []
        self._current_issue_id = ""
        self._current_snapshot_id = ""
        self._current_target_url = ""
        self._current_model = ""

        callbacks.setExtensionName("Burp AI Bridge Companion")
        callbacks.registerContextMenuFactory(self)
        callbacks.addSuiteTab(self)

        self._panel = JPanel(BorderLayout())
        mono = Font("Consolas", Font.PLAIN, 12)
        self._output = JTextArea("", 18, 100)
        self._output.setEditable(False)
        self._output.setLineWrap(False)
        self._output.setWrapStyleWord(False)
        self._output.setFont(mono)
        self._request_view = JTextArea("", 16, 100)
        self._request_view.setEditable(False)
        self._request_view.setLineWrap(False)
        self._request_view.setWrapStyleWord(False)
        self._request_view.setFont(mono)
        self._response_view = JTextArea("", 16, 100)
        self._response_view.setEditable(False)
        self._response_view.setLineWrap(False)
        self._response_view.setWrapStyleWord(False)
        self._response_view.setFont(mono)
        controls = JPanel(GridLayout(2, 1))
        toolbar = JPanel()
        info = JPanel()

        self._base_url_field = JTextField(self._base_url, 28)
        self._issue_field = JTextField("", 18)
        self._snapshot_field = JTextField("", 18)
        self._current_model_field = JTextField("", 18)
        self._switch_model_field = JTextField("", 18)
        self._issue_field.setEditable(False)
        self._snapshot_field.setEditable(False)
        self._current_model_field.setEditable(False)

        open_button = JButton("Open Plan in Repeater", actionPerformed=self._open_plan_action)
        sync_button = JButton("Sync Selected Tabs", actionPerformed=self._sync_tabs_action)
        refresh_button = JButton("Refresh Panel State", actionPerformed=self._refresh_panel_action)
        recommend_button = JButton("Show Burp Moves", actionPerformed=self._recommend_burp_moves_action)
        refresh_models_button = JButton("Refresh Model Options", actionPerformed=self._refresh_model_options_action)
        switch_model_button = JButton("Switch Active Model", actionPerformed=self._switch_model_action)
        reset_model_button = JButton("Reset Model", actionPerformed=self._reset_model_action)

        toolbar.add(JLabel("Bridge URL"))
        toolbar.add(self._base_url_field)
        toolbar.add(open_button)
        toolbar.add(sync_button)
        toolbar.add(refresh_button)
        toolbar.add(recommend_button)
        toolbar.add(refresh_models_button)
        info.add(JLabel("Issue"))
        info.add(self._issue_field)
        info.add(JLabel("Snapshot"))
        info.add(self._snapshot_field)
        info.add(JLabel("Model"))
        info.add(self._current_model_field)
        info.add(JLabel("Switch To"))
        info.add(self._switch_model_field)
        info.add(switch_model_button)
        info.add(reset_model_button)

        controls.add(toolbar)
        controls.add(info)
        self._panel.add(controls, BorderLayout.NORTH)
        output_scroll = JScrollPane(self._output)
        output_scroll.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED)
        output_scroll.setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED)
        request_scroll = JScrollPane(self._request_view)
        request_scroll.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED)
        request_scroll.setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED)
        response_scroll = JScrollPane(self._response_view)
        response_scroll.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED)
        response_scroll.setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED)

        tabs = JTabbedPane()
        tabs.addTab("Bridge Output", output_scroll)
        tabs.addTab("Selected Request", request_scroll)
        tabs.addTab("Selected Response", response_scroll)

        self._panel.add(tabs, BorderLayout.CENTER)
        self._append("Burp AI Bridge companion loaded.\n")

    def getTabCaption(self):
        return "AI Bridge"

    def getUiComponent(self):
        return self._panel

    def createMenuItems(self, invocation):
        self._selected_messages = list(invocation.getSelectedMessages() or [])
        menu = ArrayList()
        menu.add(JMenuItem("AI Bridge: Open plan in Repeater", actionPerformed=lambda event: self._open_plan_action(event)))
        menu.add(JMenuItem("AI Bridge: Sync selected tabs", actionPerformed=lambda event: self._sync_tabs_action(event)))
        menu.add(JMenuItem("AI Bridge: Refresh panel state", actionPerformed=lambda event: self._refresh_panel_action(event)))
        menu.add(JMenuItem("AI Bridge: Show Burp moves", actionPerformed=lambda event: self._recommend_burp_moves_action(event)))
        return menu

    def _open_plan_action(self, _event):
        payload = self._build_single_message_payload(source_tool="scanner")
        if not payload:
            self._append("No Burp message selected for plan open.\n")
            return
        result = self._post_json("/api/burp/open-repeater-plan?dispatch=true", payload)
        self._sync_identity(result)
        self._render_panel_state("Open plan result", result)

    def _sync_tabs_action(self, _event):
        payload = self._build_sync_payload()
        if not payload:
            self._append("Need at least one selected Repeater request/response to sync.\n")
            return
        result = self._post_json("/api/burp/repeater-sync", payload)
        self._sync_identity(result)
        self._render_panel_state("Repeater sync result", result)

    def _refresh_panel_action(self, _event):
        payload = self._build_single_message_payload(source_tool="repeater") or {
            "raw_request": "",
            "target_url": self._current_target_url,
            "use_burp_mcp_context": True,
        }
        query = []
        if self._current_issue_id:
            query.append("issue_id=" + self._current_issue_id)
        if self._current_snapshot_id:
            query.append("snapshot_id=" + self._current_snapshot_id)
        endpoint = "/api/burp/panel-state"
        if query:
            endpoint += "?" + "&".join(query)
        result = self._post_json(endpoint, payload)
        self._sync_identity(result)
        self._render_panel_state("Panel state", result)

    def _recommend_burp_moves_action(self, _event):
        payload = self._build_single_message_payload(source_tool="scanner") or {
            "raw_request": "",
            "target_url": self._current_target_url,
            "use_burp_mcp_context": True,
        }
        result = self._post_json("/api/burp/capability-recommendations", payload)
        self._render_capability_recommendations(result)

    def _refresh_model_options_action(self, _event):
        result = self._get_json("/api/runtime/model-options")
        self._sync_model_state(result)
        self._append("Model options refreshed.\n")

    def _switch_model_action(self, _event):
        model_name = self._switch_model_field.getText().strip()
        if not model_name:
            self._append("Enter a model tag before switching.\n")
            return
        result = self._post_json("/api/runtime/model-selection", {"model_name": model_name, "source": "burp-panel"})
        self._sync_model_state(result)
        self._append("Active model switched to %s.\n" % result.get("current_model", model_name))

    def _reset_model_action(self, _event):
        result = self._delete_json("/api/runtime/model-selection")
        self._sync_model_state(result)
        self._append("Active model reset to env default.\n")

    def _build_single_message_payload(self, source_tool="repeater"):
        if not self._selected_messages:
            return None
        message = self._selected_messages[0]
        request_text = self._helpers.bytesToString(message.getRequest() or "")
        response_text = self._helpers.bytesToString(message.getResponse() or "")
        self._request_view.setText(request_text)
        self._request_view.setCaretPosition(0)
        self._response_view.setText(response_text)
        self._response_view.setCaretPosition(0)
        analyzed = self._helpers.analyzeRequest(message)
        url = str(analyzed.getUrl()) if analyzed and analyzed.getUrl() else ""
        method = analyzed.getMethod() if analyzed else ""
        self._current_target_url = url or self._current_target_url
        return {
            "raw_request": request_text,
            "raw_response": response_text,
            "target_url": url,
            "http_method": method,
            "source_tool": source_tool,
            "use_burp_mcp_context": True,
        }

    def _build_sync_payload(self):
        if not self._selected_messages:
            return None

        baseline = self._helpers.bytesToString(self._selected_messages[0].getResponse() or "")
        if self._selected_messages:
            first_request = self._helpers.bytesToString(self._selected_messages[0].getRequest() or "")
            self._request_view.setText(first_request)
            self._request_view.setCaretPosition(0)
            self._response_view.setText(baseline)
            self._response_view.setCaretPosition(0)
        observations = []
        for index, message in enumerate(self._selected_messages[1:]):
            analyzed = self._helpers.analyzeRequest(message)
            observations.append({
                "tab_name": "selected-tab-%d" % (index + 1),
                "request_text": self._helpers.bytesToString(message.getRequest() or ""),
                "response_text": self._helpers.bytesToString(message.getResponse() or ""),
                "request_ref": str(analyzed.getUrl()) if analyzed and analyzed.getUrl() else "",
            })

        payload = self._build_single_message_payload(source_tool="repeater") or {}
        payload["baseline_response_text"] = baseline
        payload["repeater_variant_observations"] = observations
        return payload

    def _post_json(self, path, payload):
        self._base_url = self._base_url_field.getText().strip() or self._base_url
        url = self._base_url.rstrip("/") + path
        body = json.dumps(payload)
        request = urllib2.Request(url, body, {"Content-Type": "application/json"})
        response = urllib2.urlopen(request, timeout=20)
        return json.loads(response.read())

    def _get_json(self, path):
        self._base_url = self._base_url_field.getText().strip() or self._base_url
        url = self._base_url.rstrip("/") + path
        response = urllib2.urlopen(url, timeout=20)
        return json.loads(response.read())

    def _delete_json(self, path):
        self._base_url = self._base_url_field.getText().strip() or self._base_url
        url = self._base_url.rstrip("/") + path
        request = urllib2.Request(url, headers={"Content-Type": "application/json"})
        request.get_method = lambda: "DELETE"
        response = urllib2.urlopen(request, timeout=20)
        return json.loads(response.read())

    def _render_panel_state(self, title, data):
        lines = [title]
        if data.get("workflow_status"):
            lines.append("Workflow: %s" % data.get("workflow_status"))
        best = data.get("best_next_tab") or {}
        if best.get("tab_name"):
            lines.append("Best next tab: %s" % best.get("tab_name"))
        diff = data.get("diff_summary") or data.get("diff") or {}
        if diff.get("summary"):
            lines.append("Diff: %s" % diff.get("summary"))
        guidance = data.get("guidance_hits") or []
        if guidance:
            top = guidance[0]
            lines.append("Matched pack: %s (%s)" % (top.get("name", ""), top.get("influence_reason", "")))
        workflow = data.get("workflow") or {}
        learning = workflow.get("mutation_learning_summary") or {}
        if learning.get("summary"):
            lines.append("Mutation learning: %s" % learning.get("summary"))
        review_summary = data.get("review_dataset_summary", "")
        if review_summary:
            lines.append("Review dataset: %s" % review_summary)
        severity = data.get("severity_assessment") or {}
        calibration = severity.get("confidence_calibration") or {}
        if calibration.get("level"):
            lines.append("Confidence: %s (%.2f)" % (calibration.get("level"), calibration.get("score", 0.0)))
        notes = data.get("notes") or []
        for note in notes[:4]:
            lines.append("- %s" % note)
        capability_recs = data.get("burp_capability_recommendations") or []
        for rec in capability_recs[:3]:
            lines.append("Burp move: %s - %s" % (rec.get("tool", ""), rec.get("why", "")))
        lines.append("")
        lines.append(json.dumps(data, indent=2, sort_keys=True))
        lines.append("")
        self._append("\n".join(lines))

    def _render_capability_recommendations(self, data):
        lines = ["Burp capability recommendations"]
        lines.append("Context source: %s" % data.get("primary_context_source", ""))
        for rec in data.get("recommendations", [])[:5]:
            lines.append("%s: %s" % (rec.get("tool", ""), rec.get("why", "")))
            lines.append("  Step: %s" % rec.get("manual_step", ""))
            lines.append("  Signal: %s" % rec.get("expected_signal", ""))
            lines.append("  Stop: %s" % rec.get("stop_when", ""))
        starter_assets = data.get("starter_assets") or {}
        if starter_assets:
            lines.append("Starter assets available: %s" % ", ".join(sorted(starter_assets.keys())))
        tools = data.get("external_tool_recommendations") or []
        for tool in tools[:4]:
            lines.append("External tool: %s (%s)" % (tool.get("tool", ""), tool.get("package_hint", "")))
        lines.append("")
        lines.append(json.dumps(data, indent=2, sort_keys=True))
        lines.append("")
        self._append("\n".join(lines))

    def _sync_identity(self, data):
        self._current_issue_id = data.get("issue_id", "") or self._current_issue_id
        self._current_snapshot_id = data.get("snapshot_id", "") or self._current_snapshot_id
        self._current_target_url = data.get("target_url", "") or self._current_target_url
        self._issue_field.setText(self._current_issue_id)
        self._snapshot_field.setText(self._current_snapshot_id)

    def _sync_model_state(self, data):
        current_model = data.get("current_model", "") or self._current_model
        if current_model:
            self._current_model = current_model
            self._current_model_field.setText(current_model)
        switch_targets = data.get("switch_targets") or {}
        if not self._switch_model_field.getText().strip():
            self._switch_model_field.setText(switch_targets.get("deep", "") or current_model)

    def _append(self, text):
        def update():
            self._output.append(text + ("\n" if not text.endswith("\n") else ""))
        SwingUtilities.invokeLater(update)
